numpy/core/src/multiarray/item_selection.c

```diff
@@ -1577,6 +1577,10 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
         if (PyArray_SIZE(mps[0]) > 0) {
             *((npy_intp *)(PyArray_DATA(ret))) = 0;
         }
+
+        if (nd > 0 && check_and_adjust_axis(&axis, nd) < 0) {
+            goto fail;
+        }
         goto finish;
     }
     if (check_and_adjust_axis(&axis, nd) < 0) {
```

Comment:
If you move this check up to line 1564, you won't need to call `check_and_adjust_axis` from two different places.

Reply:
@eric-wieser 

We cant move it above the loop check for the following reason

```c
check_and_adjust_axis_msg(int *axis, int ndim, PyObject *msg_prefix)
{
    /* Check that index is valid, taking into account negative indices */
    if (NPY_UNLIKELY((*axis < -ndim) || (*axis >= ndim))) {
        /*
         * Load the exception type, if we don't already have it. Unfortunately
         * we don't have access to npy_cache_import here
         */
        static PyObject *AxisError_cls = NULL;
        PyObject *exc;

        if (AxisError_cls == NULL) {
            PyObject *mod = PyImport_ImportModule("numpy.core._exceptions");

            if (mod != NULL) {
                AxisError_cls = PyObject_GetAttrString(mod, "AxisError");
                Py_DECREF(mod);
            }
        }

        /* Invoke the AxisError constructor */
        exc = PyObject_CallFunction(AxisError_cls, "iiO",
                                    *axis, ndim, msg_prefix);
        if (exc == NULL) {
            return -1;
        }
        PyErr_SetObject(AxisError_cls, exc);
        Py_DECREF(exc);

        return -1;
    }
    /* adjust negative indices */
    if (*axis < 0) {
        *axis += ndim;
    }
    return 0;
}
```

For 0d arrays, axis checks will fail

eg:
```python
np.lexsort((np.array(0),))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in lexsort
numpy.AxisError: axis -1 is out of bounds for array of dimension 0
```

Reply:
One thing we could do is not call, axis check when no axis is mentioned i.e axis = -1

```c
    if (axis != -1 && check_and_adjust_axis(&axis, nd) < 0) {
        goto fail;
    }
```

I am however not sure, if it is something we can do. Trying to figure the implications

Reply:
We basically already do that elsewhere:

https://github.com/numpy/numpy/blob/8dfaa4a5e0ca329e16a853d5982ea5af5d044b47/numpy/core/src/multiarray/conversion_utils.c#L289-L299

Reply:
So probably soemthing like:
```C
         /* 
          * Special case letting axis={-1,0} slip through for scalars, 
          * for backwards compatibility reasons. 
          */ 
         if (ndim == 0 && (axis == 0 || axis == -1)) { 
            /* TODO: can we deprecate this? */
         }
         else if (check_and_adjust_axis(&axis, nd) < 0) {
             goto fail;
         }
```

Reply:
Thanks for pointing it out will update changes